WIP: The great pairwise reduction refactor

mbauman opened the pull request:
This is the grand culmination of (and core motivation for) all the reductions work I've slowly been chipping away at over the past year.  It implements the following major changes:

* All `reduce`-based reductions use a pairwise re-association by default.  This includes **seven** different implementations:
    * Whole array reductions
        * IndexLinear (these were often pairwise already)
        * IndexCartesian (these were previously `foldl`s without SIMD)
        * Iterators with Shape or Length (previously `foldl`s without SIMD)
        * Iterators with unknown length (previously `foldl`s without SIMD)
    * Dimensional array reductions
        * Contiguous column major dimensions (only IndexLinear were pairwise before)
        * Discontiguous column major dimensions
        * Row major dimensions
* All of these implementations do their inner work through a singular `mapreduce_kernel` function (with specific signatures for arrays and iterators) after performing the pairwise splits and/or dimensional selection.  This could eventually become a public API that should help alleviate some of the disruption to those who had been extending the internals before.  See, for example, how LinearAlgebra replaced its internal and tricky `_sum` overloads with this method: https://github.com/JuliaLang/LinearAlgebra.jl/commit/e70953e599be1d43ead342c0f56b44f008900be8.
* The dimensional array reductions no longer guess at initialization. This fully replaces my previous attempt in #55318, and is at the core of what I want to address here — the guesswork here is what was responsible for many bugs.
* All of those implementations consistently use `init` to start every reduction chain as we decided the documentation should direct implementations to do in #53945. They also consistently use `mapreduce_first`.

This currently stands atop the yet-to-be-merged #58374 and #55628.

As the successor to #55318, this carries all its tests and addresses all those issues in the dimensional reductions. Closes: https://github.com/JuliaLang/julia/issues/45566, closes https://github.com/JuliaLang/julia/issues/47231, closes https://github.com/JuliaLang/julia/issues/55213, closes https://github.com/JuliaLang/julia/issues/31427, closes https://github.com/JuliaLang/julia/issues/41054, closes https://github.com/JuliaLang/julia/issues/54920, closes https://github.com/JuliaLang/julia/issues/42259, closes https://github.com/JuliaLang/julia/issues/38660, closes https://github.com/JuliaLang/julia/issues/21097, closes https://github.com/JuliaLang/julia/issues/32366, closes https://github.com/JuliaLang/julia/issues/54875, closes https://github.com/JuliaLang/julia/issues/43731, closes https://github.com/JuliaLang/julia/issues/45562.

In addition, this further brings pairwise reassociations (and SIMD!) to the other class of reduction.  Closes https://github.com/JuliaLang/julia/issues/29883, closes https://github.com/JuliaLang/julia/issues/52365, closes https://github.com/JuliaLang/julia/issues/30421, closes https://github.com/JuliaLang/julia/issues/52457, closes https://github.com/JuliaLang/julia/issues/39385.

This requires some changes to the stdlibs:  It requires https://github.com/JuliaStats/Statistics.jl/pull/172 (which fixes another four issues: https://github.com/JuliaStats/Statistics.jl/issues/7, https://github.com/JuliaStats/Statistics.jl/issues/126, https://github.com/JuliaStats/Statistics.jl/issues/160, https://github.com/JuliaStats/Statistics.jl/issues/183). This also requires changes to SparseArrays and LinearAlgebra.

There are other pull requests that this supersedes that were inspirational for my work here.  In particular, this supersedes and would close https://github.com/JuliaLang/julia/pull/45651 (pairwise whole-array IndexCartesian), closes https://github.com/JuliaLang/julia/pull/52397 (pairwise whole-array Iterators.SizeUnknown). It also supersedes a few other PRs; would close https://github.com/JuliaLang/julia/pull/45822 (findmin/max for offset arrays; I've added those tests here), close https://github.com/JuliaLang/julia/pull/58241 (my PR for whole-array reductions `init` refactor that snowballed into this one).

Yet to do:
* [ ] Benchmark. This *should* also enable the often-requested performance benefits of SIMD reassociations over the existing foldl fallbacks, but I have yet to do consistent benchmarks.  Some specific perf issues to check include https://github.com/JuliaLang/julia/issues/33928, https://github.com/JuliaLang/julia/issues/28848, https://github.com/JuliaLang/julia/issues/47216, https://github.com/JuliaLang/julia/issues/43501.  Unfortunately it looks like this makes https://github.com/JuliaLang/julia/issues/58370 wayyyy worse.
* [x] Add more tests for other issues. I believe this will also address https://github.com/JuliaLang/julia/issues/39385.
* [ ] The initial SparseArrays branch here is the absolute minimum to get tests passing without `--depwarn=error` — all it needed was a `mapreduce_similar` definition to satisfy some type checks!~~ But because the internals changed, the sparse optimizations are no longer taking effect — they're just adding methods to vestigial deprecated methods that aren't called.
    * [x] Remove deprecations and define the basic reduction kernels
    * [ ] Check benchmarks and add more optimizations as necessary; I think this will largely be around allowing Sparse arrays to use a larger pairwise blocksize. 
* [ ] Add an opt-out for pairwise re-association that would just call the `mapreduce_kernel` on the entire dimension/array.  This would be helpful for integers and other fully associative operations
* [x] Get CI passing
    * [x] Remove the uses of the deprecations
    * [x] OffsetArrays allows for (and intentionally tests) reducing into differently-offset singleton dimensions :-\
* [ ] Run nanosoldier 😬 and evaluate its results
* [ ] More documentation and NEWS

mbauman committed:
unify reduce/reducedim empty case behaviors

Previously, Julia tried to guess at initial values for empty dimensional reductions in a slightly different way than whole-array reductions. This change unifies those behaviors, such that `mapreduce_empty` is called for empty dimensional reductions, just like it is called for empty whole-array reductions.

Beyond the unification (which is valuable in its own right), this change enables some downstream simplifications to dimensional reductions and is likely to be the "most breaking" public behavior in a refactoring targeting more correctness improvements.

(cherry picked from commit 94f24b8bddf07927eb03f84ba9ef50f966a26633)

mbauman committed:
TEMPORARILY USE SPARSEARRAYS PR BRANCH

(cherry picked from commit 0e924cf003de1673af4ca775504049636ed7ffd1)

mbauman committed:
counting and summing Bools with a small integer `init` promote to `[U]Int`.

(cherry picked from commit c923b1ddf06466ab094bf8af1ecff3c30fa187e8)

mbauman committed:
Fix doctests

(cherry picked from commit 4bee1914f17bd512e15b07c3cdab1e042471cc6d)

mbauman committed:
keep the add_sum method table small

for accumulate inference

(cherry picked from commit 4fadd8f2cbc3dee450a996936f8e037f17186980)

mbauman committed:
add two-arg axes and size for Broadcasted

mbauman committed:
fully pairwise mapreduce with consistent use of init

mbauman committed:
fixup Broadcast size/axes

mbauman committed:
more deprecations

mbauman committed:
NEWS update for empty dimensional reductions

mbauman committed:
avoid calling length in functions that define length

mbauman committed:
Add (offset) axes support for Enumerate

they already supported axes through the default fallback that uses size to construct OneTos

mbauman committed:
make it a bit easier to extend `mapreduce_kernel`

mbauman committed:
fixup PermutedDimsArray changes

mbauman committed:
avoid similar shenanigans; they are not needed

and this was causing ambiguities

mbauman committed:
cartesian broadcasts are now pairwise!

mbauman committed:
TEMPORARILY USE LINEAR ALGEBRA BRANCH

mbauman committed:
promote to Int

(cherry picked from commit 32fc16babf14e0ac26ca36adaedee4d1a3d4f0c4)

mbauman committed:
restore `reduce_first(+, true)` method

mbauman committed:
TEMPORARILY USE SPARSEARRAY AND STATISTICS BRANCHES

tecosaur commented:
Fantastic stuff, this is really exciting to see! :partying_face:

oscardssmith commented on base/dict.jl:
```diff
@@ -845,7 +845,15 @@ function iterate(d::ImmutableDict{K,V}, t=d) where {K, V}
-length(t::ImmutableDict) = count(Returns(true), t)
+# length is defined in terms of iteration; using a higher-order function to do the iteration
+# is likely to call `length` again, so this is a manual for loop:
+function length(t::ImmutableDict)
```
won't this be horribly slow?

mbauman replied:
I mean, wasn't it already doing this?  This shouldn't be any worse.  It's definitely faster than infinite recursion :)

The trouble is that `ImmutableDict` has an `IteratorSize` of `HasLength`.  And so the new reduction machinery happily calls `length` to determine the pairwise splits.  So if `length` itself uses reductions (like count) to calculate length, we're in trouble.

On nightly:

```julia
julia> using BenchmarkTools

julia> d = Base.ImmutableDict((i => i+1 for i in 1:200)...);

julia> @benchmark length($d)
BenchmarkTools.Trial: 10000 samples with 641 evaluations per sample.
 Range (min … max):  193.512 ns … 392.225 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     193.772 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   196.869 ns ±  11.655 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █▁  ▁▃▂▄▂                                                     ▁
  ███████████▇██▇▇▆▇▆▆▆▅▅▇▆▆▅▆▅▅▆▅▅▅▅▅▅▅▄▄▅▅▅▅▂▄▃▃▄▄▄▄▃▂▂▂▃▄▂▅▄ █
  194 ns        Histogram: log(frequency) by time        241 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @eval Base function length(t::ImmutableDict)
           r = 0
           for _ in t
               r+=1
           end
           return r
       end
length (generic function with 95 methods)

julia> @benchmark length($d)
BenchmarkTools.Trial: 10000 samples with 631 evaluations per sample.
 Range (min … max):  193.542 ns … 366.482 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     193.740 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   196.829 ns ±  11.533 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █   ▁ ▄ ▂                                                     ▁
  ███████▇███▇██▇▇▇▇▆▇▆▆▅▆▆▆▇▆▆▆▅▅▅▅▅▅▄▅▆▅▄▅▅▅▅▄▄▅▄▄▄▅▅▄▄▃▄▄▃▃▄ █
  194 ns        Histogram: log(frequency) by time        238 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

jakobnissen commented:
Worth seeing if it also closes https://github.com/JuliaLang/julia/issues/43501

JamesWrigley commented:
Very cool work :)

> * All of these implementations do their inner work through a singular `mapreduce_kernel` function (with specific signatures for arrays and iterators) after performing the pairwise splits and/or dimensional selection.  This could eventually become a public API that should help alleviate some of the disruption to those who had been extending the internals before.  See, for example, how LinearAlgebra replaced its internal and tricky `_sum` overloads with this method: [JuliaLang/LinearAlgebra.jl@e70953e](https://github.com/JuliaLang/LinearAlgebra.jl/commit/e70953e599be1d43ead342c0f56b44f008900be8).

This is a great idea, and I wonder if it would be possible to take it one step further and allow somehow passing the index of the input/output arrays to the function `f`? Asking because I would love to re-use this machinery to implement non-standard reductions like a single-pass `nanmean`: https://github.com/brenhinkeller/NaNStatistics.jl/issues/54#issuecomment-2852799909
There I want to do a sum into the output array but also modify a `counts` array to normalize properly. In that example I had to copy the code from Base because none of the mapreduce functions pass the index to the ouput.

giordano commented:
> Some specific perf issues to check include [...] #47216

```julia-repl
julia> f(n) = sum(i -> 4/(4i-2n-3), 1:n)
f (generic function with 1 method)

julia> g(n) = sum(i -> 4/(4i-2n-3), 1:n; init=0.0)
g (generic function with 1 method)

julia> using BenchmarkTools

julia> @btime f(100_000_000)
  44.522 ms (0 allocations: 0 bytes)
3.1415926435897923

julia> @btime g(100_000_000)
  44.544 ms (0 allocations: 0 bytes)
3.1415926435897923

julia> versioninfo()
Julia Version 1.13.0-DEV.601
Commit 6cc2496a81f (2025-05-14 21:53 UTC)
Build Info:
  Official https://julialang.org release
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 22 × Intel(R) Core(TM) Ultra 7 155H
  WORD_SIZE: 64
  LLVM: libLLVM-19.1.7 (ORCJIT, alderlake)
  GC: Built with stock GC
Threads: 1 default, 1 interactive, 1 GC (on 22 virtual cores)
```

The performance difference seems to be gone (yay!), I hope `g` doesn't have an error path which prevents static compilation though (it didn't at the time when I opened the issue, I didn't check here) :sweat_smile:

giordano commented:
> Unfortunately it looks like this makes #58370 wayyyy worse.

Confirmed the massive slowdown:
```julia-repl
julia> @b sum(f, $v), sum(f, x for x in $v)
(371.862 ns, 142.086 ns) # 1.13.0-DEV.480 Commit 89003fe49c6
(379.000 ns, 38.788 μs (1004 allocs: 15.703 KiB)) # 1.13.0-DEV.601 Commit 6cc2496a81f
```

Main problem seems to be that this is now allocating memory, which feels like it shouldn't at all, maybe it's not too hard to fix it.

jakobnissen commented on base/reduce.jl:
```diff
@@ -420,34 +416,176 @@ The default is `reduce_first(op, f(x))`.
+    v = _mapreduce_start(f, op, A, init, a1)
+    length(inds) == 1 && return v
+    @simd for i in inds[begin+1:end]
+        a = @inbounds A[i]
```
This `@inbounds` is too unsafe, IMO. It makes a wrong `length` implementation segfault. For the most basic case (e.g. iterating over a `Vector`), the compiler should be able to remove the bounds check automatically.

mbauman replied:
This is the status quo, it's just spread around everywhere currently: [1](https://github.com/JuliaLang/julia/blob/master/base/reduce.jl#L262) and [2](https://github.com/JuliaLang/julia/blob/eebda6c040f7285d2ee1eb69be8f5a7a203b43dc/base/reduce.jl#L439) and [3](https://github.com/JuliaLang/julia/blob/eebda6c040f7285d2ee1eb69be8f5a7a203b43dc/base/reducedim.jl#L281) and [4](https://github.com/JuliaLang/julia/blob/eebda6c040f7285d2ee1eb69be8f5a7a203b43dc/base/reducedim.jl#L1036) and ...

This is needed because Julia doesn't currently hoist the bounds checks in the most simple of next cases.  It's also limited to indexing into `::AbstractArray`.  It's what we demand of `AbstractArray` implementations.

jakobnissen reviewed (commented):
W.r.t the example in https://github.com/JuliaLang/julia/issues/58370, the performance difference comes from two different sources:
1. Some of the functions vectorize and others do not
2. LLVM creates different quality vectorized code depending on fickle circumstances

IMO we should not do anything about 2), since that's mostly out of our control, but we should make sure the functions are vectorizable.
The performance difference does not come from inference failure or union-splitting failure or lack of inlining or anything like that, AFAICT

oscardssmith commented on base/reduce.jl:
```diff
@@ -420,34 +416,176 @@ The default is `reduce_first(op, f(x))`.
+function mapreduce_kernel(f, op, itr, init, ::Union{HasLength, HasShape}, n, state...)
+    a1, s = iterate(itr, state...)
+    v = _mapreduce_start(f, op, itr, init, a1)
+    @simd for _ in 2:n
+        a, s = iterate(itr, s)
+        v = op(v, f(a))
```
> By using @simd, you are asserting
  several properties of the loop:
    •  It is safe to execute iterations in arbitrary or overlapping order

This isn't safe (neither the original version nor the updated one). Reordering the calls to `iterate` would be completely broken here.

mbauman replied:
This surprised me, too: https://discourse.julialang.org/t/is-it-ok-to-iterate-inside-a-simd-loop/128338

mbauman replied:
So here's a thought: what if we moved the simdability property of a reduction into the reduction function itself?  `sum` is already using a special `add_sum` method. We do similarly with several other high-level reductions... and there aren't all that many simdy operators in the first place.  Why not explicitly annotate the operator itself with `contract reassoc`?  That is:

```Julia
julia> add_reassoc_contract(x::Float32,y::Float32) = Base.llvmcall("""
           %3 = fadd reassoc contract float %0, %1
           ret float %3""", Float32, Tuple{Float32, Float32}, x, y)
[ Warning: llvmcall is defined in Core.Intrinsics and is not public in Base
add_reassoc_contract (generic function with 1 method)

julia> function foldl(op, itr, init) # the simplest of naive definitions
           (a1, s) = iterate(itr)
           r = op(init, a1)
           for a in Iterators.rest(itr, s)
               r = op(r, a)
           end
           return r
       end
foldl (generic function with 1 method)

julia> using BenchmarkTools

julia> @btime foldl(+, $(rand(Float32, 10000)), 0f0);
  9.166 μs (0 allocations: 0 bytes)

julia> @btime foldl(add_reassoc_contract, $(rand(Float32, 10000)), 0f0);
  567.478 ns (0 allocations: 0 bytes)
```

There are a few other things the `@simd` macro gives us.  There's the splitting of cartesian iterations into outer and inner loops (which I'm already manually doing), it tries to be clever about using a `while` loop (with a note that it's was optimizing for what LLVM 3.3 could do), and there's additionally annotating multiplications that then get summed as `contract reassoc` — targeting SIMD FMAs for the classic dot product.  But the advantages here aren't as big or as definitive.  On my Aarch64 M1, I do get `fmadd`s and `fmla`s in the native code, but no perf gain and I'm certain what situations you'd see a numerical improvement; this only happens with _sums_ where we're highly likely to wash away final bits in any case.

```julia
julia> mul_reassoc_contract(x::Float32,y::Float32) = Base.llvmcall("""
           %3 = fmul reassoc contract float %0, %1
           ret float %3""", Float32, Tuple{Float32, Float32}, x, y)
mul_reassoc_contract (generic function with 1 method)

julia> A = rand(Float32, 10000);

julia> @btime foldl(add_reassoc_contract, (x * x for x in $A), 0f0)
  747.262 ns (0 allocations: 0 bytes)
3259.1987f0

julia> @btime foldl(add_reassoc_contract, (mul_reassoc_contract(x, x) for x in $A), 0f0)
  746.926 ns (0 allocations: 0 bytes)
3259.1987f0
```

In any case, these weren't things that _reductions_ could even get close to optimizing in the past.  Compare to nightly's:

```Julia
julia> @btime sum((x*x for x in $A))
  9.167 μs (0 allocations: 0 bytes)

julia> @btime sum(Broadcast.broadcasted(*, $A, $A))
  9.167 μs (0 allocations: 0 bytes)
```

mbauman replied:
Ah, but that's precisely the beauty here — Julia/LLVM don't need this leeway with Ints!  The answer won't change so it'll automatically SIMD if it can.  You needn't write `@simd` (or fast math flags) to get their reductions to SIMD:

```julia
julia> function foldl(op, itr, init)
           (a1, s) = iterate(itr)
           r = op(init, a1)
           for a in Iterators.rest(itr, s)
               r = op(r, a)
           end
           return r
       end
foldl (generic function with 1 method)

julia> A = rand(Int32.(1:10), 10000);

julia> @btime foldl(+, $A, Int32(0));
  386.960 ns (0 allocations: 0 bytes)

julia> @btime foldl(+, $(Int16.(A)), Int16(0));
  191.181 ns (0 allocations: 0 bytes)
```

Unfortunately this doesn't play as nicely when we promote to Int with the builtin `Base.add_sum`, but `@simd` doesn't save us there, either.

mbauman commented:
> The performance difference does not come from inference failure or union-splitting failure or lack of inlining or anything like that, AFAICT

I don't have any other alternative explanations for the O(n) allocations that we see.  Do you?

jakobnissen commented:
You are right - construction of the `(a, s)` tuple allocates because the tuple is abstract (`a` being a union type). Rewriting the loop body to the following solves the issue:
```
it = iterate(itr, s)
v = op(v, f(first(it)))
s = last(it)
```
But you are right this is brittle.

mbauman committed:
TEMPORARILY USE SPARSEARRAYS: mapreduce_kernel implementation!

mbauman committed:
fix offset IndexLinears

mbauman committed:
support differently-offset output arrays

grumble grumble

mbauman committed:
remove not-fully-fixed MWE from 52457

mbauman committed:
implement the kernel for SCartesianIndices

mbauman commented:
OK, CI isn't near as ugly as I first thought at the quick glance at all the red.  Most of it is from usages of the deprecated things because CI runs with `--depwarn=error`.  The biggest problem is that OffsetArrays allows for in-place reductions into differently-offset arrays.  In other words, you can do `maximum!(similar(A, (42:42, axes(A, 2)), A)` and it'll happily do the `dims=2` thing, regardless of what the first axes are.  I really don't like it (despite being the one who added most of the tests for it long ago!). It's a little fiddly, but we can support it here.  It might be a good candidate for a change in the future.

Interestingly, the old SparseArray reduction machinery is overly aggressively optimized for non-commutative operators, but it's not tested and I don't think anyone has discovered/reported/cared about it.

```julia
julia> A = sparse([1,0,1,0])
4-element SparseVector{Int64, Int64} with 2 stored entries:
  [1]  =  1
  [3]  =  1

julia> mapreduce(string, *, A)
"0011"

julia> mapreduce(string, *, Array(A))
"1010"
```

This branch now fixes that, although at a relatively stiff cost (need tests though).  The biggest thing standing in the way of sparse performance right now is the fact that it'll recursively split down to 1024 elements — stored or not!  Gotta figure out a better way to determine the pairwise_blocksize.

mbauman committed:
remove the mapreduce patches from juliac buildscript

we probably need some of these function specializations, but I do not know how to find them

mbauman committed:
fix error test that was accidentally tripping over depwarns

mbauman committed:
TEMPORARILY USE LINEARALGEBRA -- bump to merge master

mbauman committed:
fix doctests

* The exact numeric results all got closer to their true values
* The Squares iterable needs length defined to sum it (as documented in the paragraph above!)

mbauman committed:
skip deprecations in doc checks

mbauman committed:
make deprecations a bit gentler

mbauman commented:
Well, I already have a pretty good sense of what's going to fail here, just by looking at packages that are already calling or extending the internals:

* Uses of [`reducedim_init`](https://juliahub.com/ui/Search?type=symbols&q=reducedim_init): these all look like method definitions that are papering over some place where the old initialization guesses were guessing wrong. I suspect none are needed
    * [ImageBase](https://github.com/JuliaImages/ImageBase.jl/blob/v0.1.7/src/utils.jl#L23)
    * [ChainRules](https://github.com/JuliaDiff/ChainRules.jl/blob/v1.72.3/src/rulesets/Base/broadcast.jl#L349)
    * [Zygote](https://github.com/FluxML/Zygote.jl/blob/v0.7.7/src/lib/broadcast.jl#L52)
* Uses of [`reducedim_initarray`](https://juliahub.com/ui/Search?type=symbols&q=reducedim_initarray). These are a tougher mix of both method definitions and uses.  Most of the definitions need to instead extend `mapreduce_similar` (as the depwarn currently advises), but I think we could support the uses with a simple fallback definition.
    * [Zygote](https://github.com/FluxML/Zygote.jl/blob/v0.7.7/src/lib/broadcast.jl#L52) and [ChainRules](https://github.com/JuliaDiff/ChainRules.jl/blob/v1.72.3/src/rulesets/Base/broadcast.jl#L355) call it in its `reducedim_init` definition from above; they shouldn't get called at all anymore, I don't think
    * [QuasiArrays](https://github.com/JuliaApproximation/QuasiArrays.jl/blob/538c66f70df20faeb9318c655af5c5b2cfd21fad/src/quasireducedim.jl#L28-L29) has a pair of definitions that I think I could just use as the deprecation's fallback
    * [DataArrays](https://github.com/JuliaStats/DataArrays.jl/blob/v0.7.0/src/reducedim.jl#L239) does something similar
    * [LazyBandedMatrices](https://github.com/JuliaLinearAlgebra/LazyBandedMatrices.jl/blob/4757c634c20156d4c54aa0074e83988c0ecf77a6/src/special.jl#L6-L9) [2](https://github.com/JuliaLinearAlgebra/LazyBandedMatrices.jl/blob/v0.11.4/src/special.jl#L6-L9) explicitly notes that they want to do something different than `similar`
    * [LazyBandedMatrices](https://github.com/JuliaLinearAlgebra/LazyBandedMatrices.jl/blob/v0.11.4/src/tridiag.jl#L598) (and more places) also calls it in their own implementations of `_sum`. They're re-implementing a lot of LinearAlgebra and could crib from what I did there — a big simplification and expansion of functionality.
    * [DistributedArrays](https://github.com/JuliaParallel/DistributedArrays.jl/blob/0451d0050c95e27d8da9a895f7cd44c49292c71a/src/mapreduce.jl#L40-L49) does something more complicated than `similar` that I don't immediately grok
    * [StatsBase](https://github.com/JuliaStats/StatsBase.jl/blob/v0.34.5/src/moments.jl#L39) calls it in a way that could be supported
* Then there's [`_mapreduce_dim`](https://juliahub.com/ui/Search?type=symbols&q=_mapreduce_dim). This is also a mix of function calls and uses.  I hadn't defined a deprecation fallback for this one, either, but I think I can.  The definitions will just miss their mark and uses can hopefully be supported seamlessly.
* There are lots of other packages that use other internals, including [`_mapreduce`](https://juliahub.com/ui/Search?type=symbols&q=_mapreduce), [`mapreducedim!`](https://juliahub.com/ui/Search?type=symbols&q=mapreducedim!), [`_mapreducedim!`](https://juliahub.com/ui/Search?type=symbols&q=_mapreducedim!), [`reducedim!`](https://juliahub.com/ui/Search?type=symbols&q=reducedim!), and [`mapreduce_impl`](https://juliahub.com/ui/Search?type=symbols&q=mapreduce_impl).  Oooof.

There's a lot here.  I fear it's gonna be a bloodbath.  I've made the deprecations slightly gentler, then let's see what Nanosoldier says...

@nanosoldier `runtests()`

quinnj commented:
Oh wow, is DataArrays.jl still a registered package? We should probably hard-deprecate/archive the repo; I can't remember the last time that was touched/updated.

mbauman commented:
Hunh, somehow the restructuring here has led to a non-deterministic flavor of `NaN` getting returned, non-deterministically.

```
Test Failed at /Users/julia/.julia/scratchspaces/a66863c6-20e8-4ff4-8a62-49f30b1f605e/agent-cache/default-grannysmith-C07ZQ07RJYVY.0/build/default-grannysmith-C07ZQ07RJYVY-0/julialang/julia-master/julia-8b8fd7c62f/share/julia/stdlib/v1.13/Statistics/test/runtests.jl:359
  Expression: var(A) === NaN
   Evaluated: NaN === NaN
```

Where `A = Int[;;]`.  I'm guessing this test should just be checking `isnan()`... and that neither we (nor LLVM) guarantee exactly which nan flavor is returned?  I'm not able to get it to trigger locally, so I can't see the exact bits.

mbauman committed:
specialize constructors in transformer for find*

N5N3 committed:
add tests from 45822

mbauman committed:
add tests for 39385

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/58418#issuecomment-2895031808) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/8b8fd7c_vs_2a5cfee/report.html) is available.

<details><summary>Report summary</summary>

## ❗ Packages that crashed

**1 packages crashed only on the current version.**

 - The process was aborted: 1 packages

6 packages crashed on the previous version too.

## ✖ Packages that failed

**1030 packages failed only on the current version.**

 - Package fails to precompile: 755 packages
 - Illegal method overwrites during precompilation: 10 packages
 - Package has test failures: 29 packages
 - Package tests unexpectedly errored: 92 packages
 - Networking-related issues were detected: 92 packages
 - There were unidentified errors: 1 packages
 - Tests became inactive: 2 packages
 - Test duration exceeded the time limit: 46 packages
 - Test log exceeded the size limit: 3 packages

1402 packages failed on the previous version too.

## ✔ Packages that passed tests

**23 packages passed tests only on the current version.**

 - Other: 23 packages

4619 packages passed tests on the previous version too.

## ~ Packages that at least loaded

**3 packages successfully loaded only on the current version.**

 - Other: 3 packages

2574 packages successfully loaded on the previous version too.

## ➖ Packages that were skipped altogether

**2 packages were skipped only on the current version.**

 - Package could not be installed: 2 packages

921 packages were skipped on the previous version too.



</details>

mbauman committed:
restore promote_union and _realtype

mbauman committed:
fixup! restore promote_union and _realtype

mbauman commented:
Yeah, I know about it but didn't think to do it here.  Would've been a good idea I suppose, even with that thousand-long-list.

-----

I did go check and confirmed that DataFrames is indeed coding up [their own `ByRow(sum)` reduction implementation](https://github.com/JuliaData/DataFrames.jl/blob/900d016e7a5f7e306b512f6633320ec1d595920d/src/abstractdataframe/selectionfast.jl#L198-L204) that is wholly independent of Base... and it is indeed a foldl.  So that failure is expected; Base is now doing better.

mbauman commented on base/reduce.jl:
```diff
@@ -420,34 +416,176 @@ The default is `reduce_first(op, f(x))`.
+_mapreduce_start(f, op, A, ::_InitialValue) = mapreduce_empty(f, op, _empty_eltype(A))
+_mapreduce_start(f, op, A, ::_InitialValue, a1) = mapreduce_first(f, op, a1)
+_mapreduce_start(f, op, A, init) = init
+_mapreduce_start(f, op, A, init, a1) = op(init, mapreduce_first(f, op, a1))
```
```suggestion
_mapreduce_start(f, op, A, init, a1) = op(init, f(a1))
```

In discussion on triage, we came to the conclusion that `mapreduce_first` exists for the case of "fixing up" the first element for use as either the return value (if its the only one) _or_ use as a first "left" argument (although we don't guarantee argument orderings).  In the case that we have `init`, we don't need to do this operation — we trust the provided `init` to effectively do it for us.

mbauman replied:
The case of adding `Bool`s gives a nice concrete example of why this is the right thing to do (and is a test-case to add).  With this suggested change, we get:

```julia
reduce(+, [true, false])     === 1    # `reduce_first(+, true)` promotes to Int, which is passed to `1 + false`
reduce(+, [true])            === 1    # that's `reduce_first(+, true)`
reduce(+, Bool[])            === 0    # that's `reduce_empty(+, Bool)`
reduce(+, [true]; init=0x00) === 0x01 # *** `reduce_first` is *not* called  and `+(::UInt8, ::Bool)` does *not* promote ***
reduce(+, Bool[]; init=0x00) === 0x00 # that's the `init`
```

This is exactly the behavior you get with `foldl` and `foldr`; they use reduce_empty and reduce_first in exactly the same way (but of course won't additionally call reduce_first a second time in a reassociated pairwise or parallel chain).  Before this suggested change the line marked `***` was out of line.

This doesn't have an impact on `sum` because `Base.add_sum(::UInt8, ::Bool)` promotes to `Int` (with https://github.com/JuliaLang/julia/pull/58374).  But that leads to an interesting question — when is it useful (or even better?) to specify promotion behaviors in a reduction operator's behavior (like `add_sum`) vs. with a cascade stemming from a promotion in `reduce_first`?  Are `add_sum`'s extra promotion behaviors even necessary if we ensure `reduce_first`'s use?

Or here's another example — Statistics' [iterable `mean` implementation](https://github.com/JuliaStats/Statistics.jl/blob/22dee82f9824d6045e87aa4b97e1d64fe6f01d8d/src/Statistics.jl#L61-L79) isn't currently written as a reduction, but it sure could be (or, really, should be). It's effectively:

```julia
add_mean(x::T, y::S) where {T,S} = x + convert(promote_type(T, S), y)
Base.reduce_first(::typeof(add_mean), x) = Base.reduce_first(+, x/1)
Base.reduce_empty(::typeof(add_mean), T) = Base.reduce_empty(+, T)/1

mean(itr) = foldl(add_mean, itr)/length(itr)
```

That initial division by one ensures a cascading promotion to float, targeting integer overflow (https://github.com/JuliaStats/Statistics.jl/pull/25).  But we also want to avoid overflow for `Float16` (https://github.com/JuliaStats/Statistics.jl/issues/140).  So what's the way to fix that?  Is it with:

```Julia
reduce_first(::typeof(add_mean), x::Float16) = Base.reduce_first(+, Float32(x)/1)
```

Or is with methods of `add_mean`? What about allowing, e..g, `FixedPoint` promotion (https://github.com/JuliaStats/Statistics.jl/issues/165)?  Should they implement methods of `add_mean` or `reduce_first`?

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/58418#issuecomment-2901432716) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/167888d_vs_2a5cfee/report.html) is available.

<details><summary>Report summary</summary>

## ❗ Packages that crashed

**1 packages crashed only on the current version.**

 - The process was aborted: 1 packages

8 packages crashed on the previous version too.

## ✖ Packages that failed

**205 packages failed only on the current version.**

 - Package fails to precompile: 11 packages
 - Illegal method overwrites during precompilation: 1 packages
 - Package has test failures: 31 packages
 - Package tests unexpectedly errored: 122 packages
 - Networking-related issues were detected: 3 packages
 - There were unidentified errors: 2 packages
 - Tests became inactive: 1 packages
 - Test duration exceeded the time limit: 32 packages
 - Test log exceeded the size limit: 2 packages

1392 packages failed on the previous version too.

## ✔ Packages that passed tests

**23 packages passed tests only on the current version.**

 - Other: 23 packages

5119 packages passed tests on the previous version too.

## ~ Packages that at least loaded

**10 packages successfully loaded only on the current version.**

 - Other: 10 packages

2907 packages successfully loaded on the previous version too.

## ➖ Packages that were skipped altogether

921 packages were skipped on the previous version too.



</details>

mbauman committed:
Specialize function args to avoid allocations

mbauman committed:
restore has_fast_linear_indexing

mbauman committed:
simplify inner indices computation

mbauman commented:
OK, getting better.  And there's still a good amount of low-hanging that should resolve more. I've fixed a few items, but some outstanding ones in the top ten listed (across the failure modes) that I haven't fully investigated or fixed or flagged upstream yet:

* There are the failures from bad (non-neutral) `init`s from #58241 (OnlineStats & AcceleratedKernels)
* Calling `norm(itr::ComponentArrays.ComponentVector{Float64, Vector{Float64}, Tuple{ComponentArrays.Axis{(p1 = 1, p2 = 2, p3 = 3, p4 = 4, p5 = 5)}}})` ends up indexing into the axes of the component vector in a way that seems to be unexpected and is leading to an ambiguity error:  `getindex(::ComponentArrays.CombinedAxis{ComponentArrays.Axis{(p1 = 1, p2 = 2, p3 = 3, p4 = 4, p5 = 5)}, Base.OneTo{Int64}}, ::UnitRange{Int64}) is ambiguous` in the `mapreduce_kernel` where we do the `inds[begin+1:end]` loop.
* ColorVectorSpace and YaoArrayRegister checks exact floating point equality in a reduction
* MatrixNetworks has some perf tests and we're 3x slower than it expects at https://github.com/JuliaGraphs/MatrixNetworks.jl/blob/ff4ea518385509b83de24e34cde5eadb430bf5f4/test/diffusions_test.jl#L227-L230. Not sure why yet
* DiskArrays and DiskArrayTools explicitly tests _how many times_ it was indexed into during a dimensional `sum` (dims=1).  It expects to be indexed into 10 times, but it's reporting 60.  This is surely because the DiskArrays.jl specialized on `Base._mapreduce` to implement some sort of chunked optimization — that's now no longer taking effect.
* Brillouin.jl gets a `-0.0` out of `cartesianize` but expects a positive zero.  The MWE there is in StaticArrays:
    ```Julia
    julia> vert = [  0.6666666666666665, -0.33333333333333326,  0.49999999999999983];

    julia> basis = [ [6.283185307179586, 3.6275987284684357, 0.0], [0.0, 7.255197456936871, 0.0], [0.0, 0.0, 5.026548245743669]];

    julia> vert' * basis
    3-element Vector{Float64}:
     4.18879020478639
     0.0
     2.5132741228718336

    julia> SVector{3}(vert)' * SVector{3}(SVector{3}.(basis))
    3-element SVector{3, Float64} with indices SOneTo(3):
      4.18879020478639
     -5.5126462140537006e-17
      2.5132741228718336
    ```
* ChainRulesCore has a `NoTangent` iterator that claims to have length (well, it's the default) but does not. Similarly for `IterativeSolvers.IDRSIterable`.
* StatsBase is picking up the wrong Statistics?
* FillArrays and BandedMatrices is an independent breakage on master (that's since been fixed, looks like)
* Clustering is a SparseArrays problem, calling `widelength` on a subarray of a sparse in the new mapreduce kernel. Should be an easy fix there.
* JuliaInterpreter expects to hit `length(LinearIndices(a))` in the `sum` implementation, which we no longer call
* [`Measurements.Derivatives`](https://github.com/JuliaPhysics/Measurements.jl/blob/ece302111abec209ce4f90bbf169bce410f7f79a/src/derivatives-type.jl#L71) uses `count` to determine its length
* Strided.jl uses `reducedim_init` in ways that aren't supported by `mapreduce_empty`.

giordano commented:
> * [Measurements.Derivatives](https://github.com/JuliaPhysics/Measurements.jl/blob/ece302111abec209ce4f90bbf169bce410f7f79a/src/derivatives-type.jl#L71) uses count to determine its length

That was based on the definition of `length(::ImmutableDict)`, which you changed here.  Could be changed there as well to mirror the new definition.

mbauman committed:
fixup! simplify inner indices computation

vtjnash commented:
StaticArrays explicitly requests the fastmath version of fma here: https://github.com/JuliaArrays/StaticArrays.jl/blob/563adebc264a23d38da97bf5d9e5d3c6333bdf26/src/matrix_multiply_add.jl#L428

adienes commented on NEWS.md:
```diff
@@ -51,6 +51,12 @@ New library features
 Standard library changes
 ------------------------
 
+* Empty dimensional reductions (`reduce`, `mapreduce`, `sum`, `minimum`, etc., with the `dims` keyword
```
It could be worth also noting in NEWS that reductions over empty axes are no longer allowed. see, this behavior on master
```
julia> sum(zeros(0, 1, 2); dims=1)
1×1×2 Array{Float64, 3}:
[:, :, 1] =
 0.0

[:, :, 2] =
 0.0

julia> sum(zeros(0, 1, 2); dims=2)
0×1×2 Array{Float64, 3}
```

and this PR makes them both errors

mbauman replied:
> PR makes them both errors

Not true! This case is ok and works _just like master_ because `sum(Float64[]) === 0.0`.  The case that's not ok is where we don't have a good `mapreduce_empty`.

```Julia
julia> Base.GIT_VERSION_INFO.commit, Base.GIT_VERSION_INFO.branch
("744e3fd976bb8a274cdca3075992d77c8712d04c", "mb/ReduceReuse♻")

julia> sum(zeros(0, 1, 2); dims=1)
1×1×2 Array{Float64, 3}:
[:, :, 1] =
 0.0

[:, :, 2] =
 0.0

julia> sum(zeros(0, 1, 2); dims=2)
0×1×2 Array{Float64, 3}

julia> sum(vcat, zeros(0, 1, 2); dims=2)
ERROR: ArgumentError: reducing over an empty collection is not allowed; consider supplying `init` to the reducer
# ...

julia> sum(vcat, Float64[])
ERROR: ArgumentError: reducing over an empty collection is not allowed; consider supplying `init` to the reducer
```

It's that second to last case that is different from master here.

mbauman replied:
(It does look like I broke something pretty wildly here with somewhere in those last few commits.)

mbauman committed:
Revert "fixup! simplify inner indices computation"

This reverts commit 744e3fd976bb8a274cdca3075992d77c8712d04c.

mbauman committed:
Revert "simplify inner indices computation"

This reverts commit be75442ac8e92a001e6ba6e09065aaa73ef339cb.

mbauman committed:
transducers for SkipMissing

mbauman committed:
fewer transducers

mbauman committed:
fixup! Specialize function args to avoid allocations

mbauman commented:
OK, I've implemented specialized kernels for `SkipMissing{<:AbstractArray}` that mirror the old implementation.  That's resolved the most egregious perf issues now; none of tested implementations are worse than a 5x regression (and quite a few are 5x+ faster).  Still haven't tested Sparse perf, though, there's definitely more work to be done there.

Here are the latest benchmarks: [144c253358.csv](https://github.com/user-attachments/files/20491072/144c253358.csv).  The remaining big regressions are in mid-size dimensional reductions, and in particular`dims=(1,2,4)`.

mbauman committed:
be more careful about which dimensions are considered "inner" reductions

mbauman committed:
skip skipmissing transducers entirely

they ended up losing type stability and it isn't hard to just define our own pairwise impl

mbauman committed:
use distinct skipmissing functions

they are just different enough to break some assumptions about this function

mbauman committed:
improve inference for unstable HasLength iters

mbauman committed:
add tuned quick outs

mbauman committed:
skip at-simd in favor of explicit re-commutating

mbauman committed:
only use mapreduce_first without initial values

and slightly improve perf for abstract iterables

mbauman committed:
simplify bootstrap; avoid vcat early in bootstrap order

mbauman committed:
Merge remote-tracking branch 'origin/master' into mb/ReduceReuse♻

mbauman committed:
fixup! Specialize function args to avoid allocations

mbauman committed:
move to 8x bins for commutative ops

mbauman committed:
Merge remote-tracking branch 'origin/master' into mb/ReduceReuse♻

mbauman commented:
OK, I must confess I've never really processed how SIMD doesn't really re-associate things.  It fundamentally _changes the entire ordering_.  It goes far above and beyond what we promise `reduce` will do... of course all the SIMD-able operations are themselves (putatively) commutative, so it's not _really_ observable outside of bit-exact floating point operations.  I think it'd be very good to completely remove `@simd` from the implementations here — or at least get rid of it from the generic ones.  Instead, we can introduce specific SIMD-like reorderings for **commutative** operations.  This can actually lead to improved performance over `@simd` in cases where the data are not stored in a data structure that directly supports a SIMD gather.  And explicitly encoding the re-ordering here would fully fix #29883 and #52457; `mean(X, dims=1)` would be exactly equal to `mean(X', dims=2)`.  And if we hard-code the re-ordering, it should completely remove architecture-dependencies from the results of these fundamental operations.

It's not uniformly a win, however.  Hard-coding such a re-ordering means that we no longer let `@simd` pick the best one for the current CPU and element type.  This PR now just picks one that works fairly well across the board.  I've definitely taken a step backwards from the previous benchmarks.  The worst offenders are the containers of `Any` and small `Union`s, and mostly where `n` (and the time itself) is very small.  In terms of pure counts, there are 5124 benchmarks in my current battery (going over various sizes, array types, iterator types, dimensions, etc.).  With https://github.com/JuliaLang/julia/pull/58418/commits/4807a128239f07c580f986367d53b657b2a6cf76 ([4807a12823.csv](https://github.com/user-attachments/files/21045193/x-4807a12823.csv)), I see:


* **0.9x runtime (or less)**: 1830 benchmarks (36%)
* **1.1x runtime (or less)**: 3196 benchmarks (62%)
* **1.5x runtime (or less)**: 4031 benchmarks (79%)
* **2x runtime (or less)**: 4579 benchmarks (89%)

A bunch of the 2x regressions are the 32-bit ones, because we're no longer saturating my M1's simd width.  And those that are much worse than that are mostly small n and/or non-concrete.  I think I can add some threshold tuning to improve some of that.

![benchmarks with concrete eltypes](https://github.com/user-attachments/assets/3bb14662-6996-40e2-aee8-ae7900a8dd69)

![benchmarks with non-concrete eltypes](https://github.com/user-attachments/assets/b68b2e8b-91fd-44b4-8294-6e254dd69a75)

adienes commented:
fixes https://github.com/JuliaStats/Statistics.jl/issues/183

adienes commented on test/reducedim.jl:
```diff
@@ -720,3 +752,345 @@ end
+some_exception(op) = try return (Some(op()), nothing); catch ex; return (nothing, ex); end
+reduced_shape(sz, dims) = ntuple(d -> d in dims ? 1 : sz[d], length(sz))
+
+@testset "Ensure that calling, e.g., sum(empty; dims) has the same behavior as sum(empty)" begin
```
fyi this testset is repeated from L579

adienes commented on base/multidimensional.jl:
```diff
@@ -1064,6 +1064,213 @@ end
+        @nexprs 8 N->begin
+            it = iterate(itr, s)
+            if it === nothing
+                N > 7 && (v_7 = op(v_7, f(a_7)))
```
```suggestion
                @nexprs 7 M->((N > M) && (v_{8-M} = op(v_{8-M}, f(a_{8-M}))))
```

adienes commented on base/multidimensional.jl:
```diff
@@ -1064,6 +1064,213 @@ end
+
+# This special internal method must have at least 8 indices and allows passing
+# optional scalar leading and trailing dimensions
+function _mapreduce_kernel_commutative(f, op, A, init, inds, leading=(), trailing=())
```
potentially profitable to unroll it out to `@nexprs 16` when on a 64bit platform but `sizeof(T) <= 4`

adienes commented on base/deprecated.jl:
```diff
@@ -558,3 +559,29 @@ true
+
+@deprecate (reducedim_initarray(A::Union{Base.AbstractBroadcasted, AbstractArray}, region, init, ::Type{T}) where {T}) fill!(mapreduce_similar(A,T,reduced_indices(A,region)), init) false
+@deprecate reducedim_initarray(A::Union{Base.AbstractBroadcasted, AbstractArray}, region, init) fill!(mapreduce_similar(A,typeof(init),reduced_indices(A,region)), init) false
+const _dep_message_reducedim_init = ", these internals have been removed. To customize the array returned by dimensional reductions, implement mapreduce_similar instead"
```
```suggestion
```

dup. on L567

adienes commented on base/reduce.jl:
```diff
@@ -302,8 +269,40 @@ implementations may reuse the return value of `f` for elements that appear multi
+mapreduce(f::F, op::G, A::AbstractArrayOrBroadcasted; init=_InitialValue(), dims=(:)) where {F,G} = mapreducedim(f, op, A, init, dims)
+mapreduce(f::F, op::G, A::AbstractArrayOrBroadcasted, As::AbstractArrayOrBroadcasted...; init=_InitialValue(), dims=(:)) where {F,G} =
+    reduce(op, map(f, A, As...); init, dims)
+mapreducedim(f::F, op::G, A, init, ::Colon) where {F,G} = mapreduce_pairwise(f, op, A, init)
```
I wonder (suspect?) if smaller `pairwise_blocksize` is better when the cost of `f` exceeds the cost of `op`? e.g. to start with:
```
pairwise_blocksize(::typeof(sqrt), op) = 64
pairwise_blocksize(::typeof(sin), op) = 64
pairwise_blocksize(::typeof(cos), op) = 64
pairwise_blocksize(::typeof(exp), op) = 64
pairwise_blocksize(::typeof(log), op) = 64
```

adienes commented on base/reduce.jl:
```diff
@@ -1109,3 +1309,8 @@ for (fred, f) in ((maximum, max), (minimum, min), (sum, add_sum))
+
+# Reduction kernels that explicitly encode simplistic SIMD-ish reorderings
+const CommutativeOps = Union{typeof(+),typeof(Base.add_sum),typeof(min),typeof(max),typeof(Base._extrema_rf),typeof(|),typeof(&)}
+is_commutative_op(::CommutativeOps) = true
```
maybe two arg is better with a default?
```
is_commutative_op(op, ::Type) = is_commutative_op(op)
```

so we can say like `*` on `Float64` commutes (for the purposes of `reduce`) but not on `Matrix` or quaternions or whatever

LilithHafner commented:
Benchmarks LGTM. It's not ideal to have some regressions come with the improvements but that's not blocking given the correctness improvements here.
